app/upgrades/v17/upgrades.go

```diff
@@ -124,12 +84,208 @@ func CreateUpgradeHandler(
+		if err != nil {
+			return err
+		}
+
+		err = authorizeSuperfluidIfEnabled(ctx, keepers, assetPair.LinkedClassicPool, clPoolDenom)
+		if err != nil {
+			return err
+		}
+
+		err = manuallySetTWAPRecords(ctx, keepers, clPoolId)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// testnetUpgradeHandler creates CL pools for all existing balancer pools. It also links the CL pools to their balancer pool counterpart, creates a full range position with the community pool,
+// authorizes superfluid for the CL pool if the balancer pool is superfluid enabled, and manually sets the TWAP records for the CL pool.
+func testnetUpgradeHandler(ctx sdk.Context, keepers *keepers.AppKeepers, communityPoolAddress sdk.AccAddress, poolLinks *[]gammmigration.BalancerToConcentratedPoolLink, fullRangeCoinsUsed *sdk.Coins) error {
+	// Retrieve all GAMM pools on the testnet.
+	pools, err := keepers.GAMMKeeper.GetPools(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, pool := range pools {
+		skipPool, gammPoolId, baseAsset, spreadFactor, err := testnetParsePoolRecord(ctx, pool, keepers)
+		if err != nil {
+			return err
+		}
+		if skipPool {
+			continue
+		}
+
+		clPoolDenom, clPoolId, err := createCLPoolWithCommunityPoolPosition(ctx, keepers, gammPoolId, baseAsset, spreadFactor, communityPoolAddress, poolLinks, fullRangeCoinsUsed)
+		if err != nil {
+			return err
+		}
+
+		err = authorizeSuperfluidIfEnabled(ctx, keepers, gammPoolId, clPoolDenom)
+		if err != nil {
+			return err
+		}
+
+		err = manuallySetTWAPRecords(ctx, keepers, clPoolId)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// createCLPoolWithCommunityPoolPosition creates a CL pool for a given balancer pool and adds a full range position with the community pool.
+func createCLPoolWithCommunityPoolPosition(ctx sdk.Context, keepers *keepers.AppKeepers, gammPoolId uint64, baseAsset string, spreadFactor sdk.Dec, communityPoolAddress sdk.AccAddress, poolLinks *[]gammmigration.BalancerToConcentratedPoolLink, fullRangeCoinsUsed *sdk.Coins) (clPoolDenom string, clPoolId uint64, err error) {
+	// Create a concentrated liquidity pool for asset pair.
+	ctx.Logger().Info(fmt.Sprintf("Creating CL pool from poolID (%d), baseAsset (%s), spreadFactor (%s), tickSpacing (%d)", gammPoolId, baseAsset, spreadFactor, TickSpacing))
+	clPool, err := keepers.GAMMKeeper.CreateConcentratedPoolFromCFMM(ctx, gammPoolId, baseAsset, spreadFactor, TickSpacing)
+	if err != nil {
+		return "", 0, err
+	}
+	clPoolId = clPool.GetId()
+	clPoolDenom = cltypes.GetConcentratedLockupDenomFromPoolId(clPoolId)
+
+	// Add the pool link to the list of pool links (we set them all at once later)
+	*poolLinks = append(*poolLinks, gammmigration.BalancerToConcentratedPoolLink{
+		BalancerPoolId: gammPoolId,
+		ClPoolId:       clPoolId,
+	})
+
+	// Get community pool balance before swap and position creation
+	commPoolBalancePre := keepers.BankKeeper.GetAllBalances(ctx, communityPoolAddress)
```

Comment:
Lets just get the osmo balance, rather than all balances. All balances can adversarally take very long.

Reply:
What I will do is get both the osmo balance and the base asset balance then. If there is dust left over of base asset (there shouldn't be but if) then we won't update fee pool properly which would panic

Reply:
Maybe its cleaner to just do a community pool spend of however many osmo to a new address, and then do all these operations of that different address? De-risks any fee pool things

Reply:
Yeah thats fine, I can change this

Reply:
@ValarDragon can you see if this sufficiently derisks it while keeping the owner as the community pool? https://github.com/osmosis-labs/osmosis/pull/5959/commits/bdb5741f64e858e7c8303ee349c9c1d6b181d22f

I believe the signaling prop was explicit about the owner of the position being the community pool, and I don't see room for too many edge cases here.

Reply:
That works then if we don't have flexibility to change addresses!